src/Agent.Worker/StepsRunner.cs

```diff
@@ -49,25 +48,32 @@ public sealed class StepsRunner : AgentService, IStepsRunner
             jobContext.Variables.Agent_JobStatus = TaskResult.Succeeded;
             foreach (IStep step in steps)
             {
-                Trace.Info($"Processing step: DisplayName='{step.DisplayName}', AlwaysRun={step.AlwaysRun}, ContinueOnError={step.ContinueOnError}, Critical={step.Critical}, Enabled={step.Enabled}, Finally={step.Finally}");
+                Trace.Info($"Processing step: DisplayName='{step.DisplayName}', ContinueOnError={step.ContinueOnError}, Critical={step.Critical}, Enabled={step.Enabled}, Finally={step.Finally}");
                 ArgUtil.Equal(true, step.Enabled, nameof(step.Enabled));
                 ArgUtil.NotNull(step.ExecutionContext, nameof(step.ExecutionContext));
                 ArgUtil.NotNull(step.ExecutionContext.Variables, nameof(step.ExecutionContext.Variables));
 
                 jobContext.Progress(stepCount++ * 100 / steps.Count);
 
-                // TODO: Run finally even if canceled?
-
-                // Skip if a previous step failed and the current step is not AlwaysRun.
-                if ((stepFailed && !step.AlwaysRun && !step.Finally)
-                    // Or if a previous Critical step failed and the current step is not Finally.
-                    || (criticalStepFailed && !step.Finally))
+                // Test critical failure.
+                if (criticalStepFailed && !step.Finally)
                 {
-                    Trace.Info("Skipping step.");
-                    
-                    // Skip the step context.
+                    Trace.Info("Skipping step due to previous critical step failure.");
                     step.ExecutionContext.Skip();
+                    continue;
+                }
 
+                // Variable expansion.
+                List<string> expansionWarnings;
+                step.ExecutionContext.Variables.RecalculateExpanded(out expansionWarnings);
+                expansionWarnings?.ForEach(x => step.ExecutionContext.Warning(x));
+
+                // Condition.
```

Comment:
I also need to figure out how to make subsequent tasks run on cancel. Currently cancel bubbles out of steps runner.

I need to spend some time to figure out the correct thing.